File: code/github_chinese1/decision_tree.cpp
#include "data.h"
#include "decision_tree.h"

using namespace std;
    
void DecisionTree::build_tree(shared_ptr<DecisionTree> tree, FeaturesLabels& dataset)
{
    Data features = dataset.features;
    vector<float> labels = dataset.labels;
    if (features.size() == 0)
    {
        return;
    }
    SplitResults feature_and_val = choose_best_feature(tree, features, labels);
    tree -> split_feature = feature_and_val.feature;
    tree -> split_value = feature_and_val.value;
    if (tree -> is_leaf == true || tree -> split_feature == -1)
    {
        return;
    }
    SplitData new_dataset = split_dataset(features, labels, tree -> split_feature, tree -> split_value);
    Data& features_1 = new_dataset.dataset_1.features;
    vector<float> labels_1 = new_dataset.dataset_1.labels;
    Data& features_2 = new_dataset.dataset_2.features;
    vector<float> labels_2 = new_dataset.dataset_2.labels;

    // 此处因为是在类/函数内定义变量，必须要分配动态内存。下面两行是不用智能指针的写法，但new需和delete配合。
    // tree -> left = new DecisionTree(tree -> height + 1);
    // tree -> right = new DecisionTree(tree -> height + 1);

    // 利用智能指针后，需要将类/函数的接口也都统一一下。
    tree -> left = make_shared<DecisionTree>(tree -> height + 1);
    tree -> right = make_shared<DecisionTree>(tree -> height + 1);

    build_tree(tree -> left, new_dataset.dataset_1);
    build_tree(tree -> right, new_dataset.dataset_2);
    return;
}

SplitResults DecisionTree::choose_best_feature(shared_ptr<DecisionTree> tree, Data& features, vector<float>& labels){

    SplitResults result;

    // 先判断下数据集是否已经完全同化
    bool flag = true;
    float value = labels.front();
    for (size_t i=0; i < features.size(); ++i)
    {
        if (labels[i] != value)
        {
            flag = false;
            break;
        }
    }
    if (flag == true || tree -> height >= tree -> maximum_height)
    {
        result.feature = -1;
        result.value = compute_mean(labels);
        tree -> is_leaf = true;
        return result;
    }
    float loss = compute_loss(features, labels);
    float maximum_gain = 0;
    for (size_t i = 0; i < features[0].size(); ++i)
    {
        for (size_t j = 0; j < features.size(); ++j)
        {
            SplitData new_dataset = split_dataset(features, labels, i, features[j][i]);
            Data& features_1 = new_dataset.dataset_1.features;
            vector<float> labels_1 = new_dataset.dataset_1.labels;
            Data& features_2 = new_dataset.dataset_2.features;
            vector<float> labels_2 = new_dataset.dataset_2.labels;
            float new_loss = compute_loss(features_1, labels_1) + compute_loss(features_2, labels_2);
            float gain = loss - new_loss;
            if (gain < tree -> threshold)
            {
                continue;
            }
            if (gain > maximum_gain)
            {
                result.feature = i;
                result.value = features[j][i];
                maximum_gain = gain;
            }
        }
    }
    if (maximum_gain == 0 || maximum_gain < tree -> threshold)
    {
        result.feature = -1;
        result.value = compute_mean(labels);
        tree -> is_leaf = true;
    }
    return result;
}

SplitData DecisionTree::split_dataset(Data& features, vector<float>& labels, int f_index, float value){
    SplitData new_dataset;
    for (size_t i = 0; i < features.size(); ++i)
    {
        if (features[i][f_index] <= value)
        {
            new_dataset.dataset_1.features.push_back(features[i]);
            new_dataset.dataset_1.labels.push_back(labels[i]);
        }
        else
        {
            new_dataset.dataset_2.features.push_back(features[i]);
            new_dataset.dataset_2.labels.push_back(labels[i]);
        }
    }
    return new_dataset;
}

float DecisionTree::compute_mean(vector<float>& labels){
    float total = 0;
    for (size_t i = 0; i < labels.size(); ++i)
    {
        total += labels[i];
    }
    return total / labels.size();
}

float DecisionTree::compute_loss(Data& features, vector<float>& labels){
    float loss = 0;
    float mean = compute_mean(labels);
    for (size_t i = 0; i < features.size(); ++i)
    {
        loss += (labels[i] - mean) * (labels[i] - mean);
    }
    return loss;
}

float DecisionTree::predict(shared_ptr<DecisionTree> tree, vector<float>& example)
{
    if (tree -> is_leaf == true || tree -> split_feature == -1)
    {
        return tree -> split_value;
    }
    if (example[tree -> split_feature] <= tree -> split_value)
    {
        return predict(tree -> left, example);
    }
    else
    {
        return predict(tree -> right, example);
    }
}

// DecisionTree的单独使用与测试方法：
// int main()
// {
//     Data* data = LoadData("/Users/liushihao/Desktop/搜索引擎基础/GBDT/bikeSpeedVsIq_train.txt");
//     FeaturesLabels training_set = split_features_labels(data);
//     DecisionTree tree;
//     tree.build_tree(&tree, training_set);

//     Data* test_set = LoadData("/Users/liushihao/Desktop/搜索引擎基础/GBDT/bikeSpeedVsIq_test.txt");
//     float loss = 0;
//     for (size_t i = 0; i < (*test_set).size(); ++i)
//     {   
//         float prediction = tree.predict(&tree, (*test_set)[i]);
//         loss += pow((prediction - (*test_set)[i].back()), 2); 
//         cout << (*test_set)[i].back() << '\t' << prediction << endl;
//     }
//     cout << "loss is " << loss;
//     return 0;
// }

File: code/stuff/github_gbdts/github_chinese1/decision_tree.py
'''
Created on Nov 30st, 2019
Tree-Based Regression Methods
@author: Shihao Liu
'''

import numpy as np

class DecisionTree:
    def __init__(self, is_leaf=False, left=None, right=None, height=1, split_feature=-1,
                 split_value=0, leaf_nums=0, threshold=0, maximum_height=10):
        self.left = left
        self.right = right
        self.height = height
        self.is_leaf = is_leaf
        self.split_feature = split_feature
        self.split_value = split_value
        self.leaf_nums = leaf_nums
        self.threshold = threshold
        self.maximum_height = maximum_height

    def build_tree(self, tree, dataset):
        if len(dataset) == 0:
            return
        best_split_feature, split_value = self.choose_best_feature(tree, dataset)
        tree.split_feature = best_split_feature
        tree.split_value = split_value
        if tree.is_leaf is True or best_split_feature == -1:
            return
        dataset_1, dataset_2 = self.split_dataset(dataset, best_split_feature, split_value)
        tree.left = DecisionTree(height=tree.height+1)
        tree.right = DecisionTree(height=tree.height+1)

        self.build_tree(tree.left, dataset_1)
        self.build_tree(tree.right, dataset_2)

    def split_dataset(self, dataset, feature, value):
        dataset_1 = dataset[np.nonzero(dataset[:, feature] <= value)[0]]
        dataset_2 = dataset[np.nonzero(dataset[:, feature] > value)[0]]
        return dataset_1, dataset_2

    def choose_best_feature(self, tree, dataset):
        if len(set(dataset[:, -1])) <= 2 or tree.height >= tree.maximum_height:
            tree.is_leaf = True
            return None, np.mean(dataset[:, -1])
        maximum_gain = 0
        feature_nums = len(dataset[0]) - 1 # 最后一列默认为label
        loss = self.compute_loss(dataset)
        for f in range(feature_nums):
            for val in dataset[:, f]:
                dataset_1, dataset_2 = self.split_dataset(dataset, f, val)
                loss_1 = self.compute_loss(dataset_1)
                loss_2 = self.compute_loss(dataset_2)
                new_loss = loss_1 + loss_2
                gain = loss - new_loss
                if gain < tree.threshold:
                    continue
                if gain > maximum_gain:
                    best_split_feature, split_value = f, val
                    maximum_gain = gain
        if maximum_gain == 0 or maximum_gain < tree.threshold:
            tree.is_leaf = True
            return None, np.mean(dataset[:, -1])
        else:
            return best_split_feature, split_value

    def compute_loss(self, dataset):
         # cart_regression
        return  np.var(dataset[:, -1]) * dataset.shape[0]

    def predict(self, tree, feature):
        if tree.is_leaf:
            return tree.split_value
        if feature[tree.split_feature] <= tree.split_value:
            return self.predict(tree.left, feature)
        else:
            return self.predict(tree.right, feature)
            
    def tree_illustration(self, tree):
        queue = [tree]
        while queue:
            tree = queue.pop(0)
            if tree.is_leaf or not type(tree.split_feature) == int:
                print(tree.split_value)
                continue
            print(tree.split_feature, tree.split_value)
            queue.extend([tree.left, tree.right])

        return

class Data:
    def __init__(self, file_path):
        self.file_path = file_path

    def load_file(self):
        dataset = np.loadtxt(self.file_path)
        return dataset

if __name__ == '__main__':
    data = Data('./bikeSpeedVsIq_train.txt')
    training_set = data.load_file()
    cart = DecisionTree()
    cart.build_tree(cart, training_set)

    # cart.tree_illustration(cart)

    data = Data('./bikeSpeedVsIq_test.txt')
    test_set = data.load_file()
    loss = 0
    for example in test_set:
        prediction = cart.predict(cart, example)
        loss += (prediction - example[-1]) ** 2

    print("模型在test_set上的最终MSE值为{:.2f}".format(loss))

    # 波士顿房价预测
    # data = load_boston()
    # X_train, X_test, y_train, y_test = train_test_split(data["data"],data["target"],test_size=0.3, random_state=0)
    # cart = DecisionTree()
    # data = np.concatenate((X_train, y_train[:, None]), axis=1)
    # cart.build_tree(cart, data)
    # data = np.concatenate((X_test, y_test[:, None]), axis=1)
    # loss = 0
    # for example in data:
    #     prediction = cart.predict(cart, example)
    #     loss += (prediction - example[-1]) ** 2
    # print("模型在test_set上的最终MSE值为{:.2f}".format(loss))
    